### PR TITLE
units: make locale directory writable for systemd-localed

### LINKED ISSUE #19138: localed's locale-gen integration does not work (Arch Linux)
systemd 248rc4 compiled with `-Dlocalegen-path=/usr/bin/locale-gen`, Arch Linux.

```
# localectl set-locale en_GB.UTF-8
Failed to issue method call: Failed to generate locale: Protocol error
```

```
Mar 27 16:38:49 arch systemd-localed[306]: rm: cannot remove '/usr/lib/locale/locale-archive': Read-only file system
Mar 27 16:38:49 arch systemd-localed[305]: Generating locales...
Mar 27 16:38:51 arch systemd[1]: systemd-udevd.service: Got notification message from PID 183 (WATCHDOG=1)
Mar 27 16:38:54 arch systemd-localed[305]:   en_US.UTF-8...
Mar 27 16:38:54 arch systemd-localed[316]: cannot open locale archive "/usr/lib/locale/locale-archive": Read-only file system
Mar 27 16:38:54 arch systemd[1]: Received SIGCHLD from PID 317 (gzip).
Mar 27 16:38:54 arch systemd[1]: Child 317 (gzip) died (code=exited, status=0/SUCCESS)
Mar 27 16:38:54 arch systemd[1]: systemd-localed.service: Child 317 belongs to systemd-localed.service.
Mar 27 16:38:54 arch systemd-localed[304]: (sd-localegen) failed with exit status 1.
Mar 27 16:38:54 arch systemd-localed[304]: Failed to generate locale: Protocol error
```

Arch's `/usr/bin/locale-gen`:

```
#!/bin/sh

set -e

LOCALEGEN=/etc/locale.gen
LOCALES=/usr/share/i18n/locales
if [ -n "$POSIXLY_CORRECT" ]; then
  unset POSIXLY_CORRECT
fi


[ -f $LOCALEGEN -a -s $LOCALEGEN ] || exit 0;

# Remove all old locale dir and locale-archive before generating new
# locale data.
rm -rf /usr/lib/locale/* || true

umask 022

is_entry_ok() {
  if [ -n "$locale" -a -n "$charset" ] ; then
    true
  else
    echo "error: Bad entry '$locale $charset'"
    false
  fi
}

echo "Generating locales..."
while read locale charset; do \
        case $locale in \#*) continue;; "") continue;; esac; \
        is_entry_ok || continue
        echo -n "  `echo $locale | sed 's/\([^.\@]*\).*/\1/'`"; \
        echo -n ".$charset"; \
        echo -n `echo $locale | sed 's/\([^\@]*\)\(\@.*\)*/\2/'`; \
        echo -n '...'; \
        if [ -f $LOCALES/$locale ]; then input=$locale; else \
        input=`echo $locale | sed 's/\([^.]*\)[^@]*\(.*\)/\1\2/'`; fi; \
        localedef -i $input -c -f $charset -A /usr/share/locale/locale.alias $locale; \
        echo ' done'; \
done < $LOCALEGEN
echo "Generation complete."
```

### PR DESCRIPTION
**Pull request by eworm-de**
With 8f20232fcb52dbe6255f3df6101fc057af90bcfa systemd-localed supports
generating locales when required. This fails if the locale directory is
read-only, so make it writable.

Closes #19138

### CODE REVIEW

**Review by bluca — APPROVED** [review]
LGTM

**Comment by bluca** [comment]
It's a new feature in 248, so adding to milestone

**Comment by poettering** [comment]
Hmpf. This looks like a dropin distro packages should add in. This should not exist on distros that dont need this.

**Comment by bluca** [comment]
> Hmpf. This looks like a dropin distro packages should add in. This should not exist on distros that dont need this.

But it's something we support upstream, no? via `-Dlocalegen-path`

**Comment by bluca** [comment]
> Hmpf. This looks like a dropin distro packages should add in. This should not exist on distros that dont need this.

Maybe make this optional based on whether `-Dlocalegen-path` is configured then? systemd-localed.service.in is already templated anyway, the change should be minimal.

**Comment by bluca** [comment]
This doesn't seem unrelated, althought probably a pre-existing system issue that appears due to enablement:

```
autopkgtest [11:47:29]: test localed-locale: [-----------------------
Generating locales (this might take a while)...
  en_US.UTF-8... done
Generation complete.
Could not get properties: Failed to activate service 'org.freedesktop.locale1': timed out (service_start_timeout=25000ms)
FAIL: 'System Locale:' not found in:

autopkgtest [11:47:55]: test localed-locale: -----------------------]
autopkgtest [11:47:55]: test localed-locale:  - - - - - - - - - - results - - - - - - - - - -
localed-locale       FAIL non-zero exit status 1
autopkgtest [11:47:56]: test localed-locale:  - - - - - - - - - - stderr - - - - - - - - - -
Could not get properties: Failed to activate service 'org.freedesktop.locale1': timed out (service_start_timeout=25000ms)
FAIL: 'System Locale:' not found in:

```

**Comment by eworm-de** [comment]
Updated to add the directory only when required.

File: meson.build [hunk 1]
```diff
@@ -860,6 +860,8 @@ have = false
 if localegen_path != ''
         conf.set_quoted('LOCALEGEN_PATH', localegen_path)
         have = true
+        locale_path_writeable = ',' + libdir + '/locale'
```
Comment (mbiebl) [inline meson.build]: Shouldn't this be a space separated list?

Comment (eworm-de) [reply meson.build]: Ah, right... Stupid me.

File: meson.build [hunk 2]
```diff
@@ -860,6 +860,8 @@ have = false
 if localegen_path != ''
         conf.set_quoted('LOCALEGEN_PATH', localegen_path)
         have = true
+        locale_path_writeable = ' ' + libdir + '/locale'
+        substs.set('SERVICE_LOCALE_PATH_WRITABLE', locale_path_writeable)
```
Comment (mbiebl) [inline meson.build]: maybe name this SERVICE_LOCALEGEN_PATH_WRITABLE or LOCALEGEN_PATH_WRITABLE. This makes the intent a bit clearer.

Comment (mbiebl) [reply meson.build]: I wonder if it wouldn't be better to just hard-code the path to `/usr/lib/locale/`. After all, this path is not really dependent on how $libdir was set during configuration.

Comment (mbiebl) [reply meson.build]: Case in point: libdir on Debian is `/usr/lib/$(DEB_HOST_MULTIARCH)`

Comment (eworm-de) [reply meson.build]: Getting it right for everybody may be anywhere between hard and impossible... How about adding a new configuration option?

Comment (mbiebl) [reply meson.build]: Atm this feature is used by Debian based distros and Arch, afaics. Both use /usr/lib/locale. In the interest of simplicity, I'd hard-code that path and only make this configurable when the need arises.

Comment (eworm-de) [reply meson.build]: Ok, fine then. :)

**Comment by eworm-de** [comment]
Oh, meson gives a warning at configuration time...

**Comment by eworm-de** [comment]
Should work now without issues and path is hardcoded.

**Pull request merged**